agent/src/main/java/com/teamscale/jacoco/agent/upload/HttpZipUploaderBase.java

```diff
@@ -32,20 +33,33 @@
 	/** Additional files to include in the uploaded zip. */
 	protected final List<Path> additionalMetaDataFiles;
 
+	/** The API class. */
+	private final Class<T> apiClass;
+
 	/** The API which performs the upload */
-	protected final T api;
+	private T api;
 
 	/** Constructor. */
-	public HttpZipUploaderBase(HttpUrl uploadUrl, List<Path> additionalMetaDataFiles) {
+	public HttpZipUploaderBase(HttpUrl uploadUrl, List<Path> additionalMetaDataFiles, Class<T> apiClass) {
 		this.uploadUrl = uploadUrl;
 		this.additionalMetaDataFiles = additionalMetaDataFiles;
+		this.apiClass = apiClass;
+	}
 
-		Retrofit retrofit = HttpUtils.createRetrofit(retrofitBuilder -> retrofitBuilder.baseUrl(uploadUrl));
-		api = getApi(retrofit);
+	/** Template method to configure the OkHttp Client. */
+	protected void configureOkHttp(OkHttpClient.Builder builder) {
 	}
 
 	/** Returns the API for creating request to the http uploader */
-	protected abstract T getApi(Retrofit retrofit);
+	protected T getApi() {
```

Comment:
Needed to change to lazy-evaluation in order to allow better customization (::configureOkHttp)